R/read-write.R

```diff
@@ -0,0 +1,465 @@
+#' Read/write a local dm from/to files
+#'
+#' @description
+#' `r lifecycle::badge("experimental")`
+#'
+#' Store a local `dm` in a file or directory and retrieve it later.
+#'
+#' @inheritParams dm_add_pk
+#' @param csv_directory For `dm_write_csv()`: The path to a non-existent or empty
+#' directory to write the `csv`-files defining the `dm` to.
+#'
+#' For `dm_read_csv()`: The path to the directory containing the `dm` as `csv-files`.
+#' @details
+#' Not all column types are supported, only: `character`, `Date`, `integer`, `logical`, `numeric`,
+#' `POSIXct`, `POSIXlt`.
+#'
+#' `POSIXct`, `POSIXlt` will be converted to timezone `UTC` to avoid incorrect results
+#' due to not tracking the timezone.
+#' An informative message will be issued when this happens.
+#'
+#' `dm_write_csv()`: write a `dm` to a collection of `csv`-files to a non-existent or empty directory.
+#'
+#' @return `read`-family: A `dm` object.
+#'
+#' `write`-family: The `dm`, invisibly.
+#'
+#' @rdname dm-read-write
+#' @export
+#' @examples
+#' \dontrun{
+#' dm_write_csv(dm_nycflights13(), "nyc_dm_as_csv")
+#' dm_read_csv("nyc_dm_as_csv")
+#'
+#' dm_write_zip(dm_nycflights13(), "dm.zip")
+#' dm_read_zip("dm.zip")
+#'
+#' dm_write_xlsx(dm_nycflights13(), "dm.xlsx")
+#' dm_read_xlsx("dm.xlsx")
+#' }
+dm_write_csv <- function(dm, csv_directory) {
+  if (dir.exists(csv_directory) && length(list.files(csv_directory)) > 0) {
+    abort_dir_not_empty()
+  }
+  tryCatch(
+    dm_write_csv_impl(dm, csv_directory, zip = FALSE),
+    error = function(e) {
+      # remove directory in case it was created but there was an error
+      try(unlink(csv_directory, recursive = TRUE))
+      # keep error message and class though
+      abort(conditionMessage(e), .subclass = class(e))
```

Comment:
The following will retain all other attributes:

```suggestion
      stop(e)
```